File: app/src/main/java/com/defianttech/convertme/WidgetSetupActivity.java
package com.defianttech.convertme;

import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.text.TextUtils;
import android.util.Log;
import android.view.MenuItem;
import android.view.View;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.Spinner;
import android.widget.SpinnerAdapter;

/*
 * Copyright (c) 2014-2016 Dmitry Brant
 */
public class WidgetSetupActivity extends AppCompatActivity {
    private static final String TAG = "WidgetSetupActivity";

    UnitCollection[] collections;
    private String[] allCategoryNames;
    private int widgetId = -1;

    private Spinner unitCategorySpinner;
    private Spinner unitFromSpinner;
    private Spinner unitToSpinner;

    private WidgetPrefs prefs;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.widget_setup_activity);

        Toolbar toolbar = (Toolbar) findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);
        getSupportActionBar().setTitle(R.string.configure_widget);

        if (!TextUtils.isEmpty(getIntent().getAction())
                && getIntent().getAction().contains(WidgetProvider.CLICK_ACTION_SETTINGS)) {
            widgetId = WidgetProvider.getWidgetId(getIntent().getAction());
        }
        if (widgetId == -1) {
            return;
        }
        Log.d(TAG, "Configuring widget " + widgetId);

        prefs = new WidgetPrefs(this, widgetId);

        collections = UnitCollection.getInstance(this);
        allCategoryNames = UnitCollection.getAllCategoryNames(this);
        SpinnerAdapter categoryAdapter = new ArrayAdapter(this, R.layout.unit_categoryitem, allCategoryNames);

        unitCategorySpinner = (Spinner) findViewById(R.id.unit_category_spinner);
        unitCategorySpinner.setAdapter(categoryAdapter);
        unitCategorySpinner.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
            @Override
            public void onItemSelected(AdapterView<?> adapterView, View view, int i, long l) {
                prefs.currentCategory = UnitCollection.collectionIndexByName(collections, allCategoryNames[i]);
                if (prefs.currentFromIndex > collections[prefs.currentCategory].length()) {
                    prefs.currentFromIndex = 0;
                }
                if (prefs.currentToIndex > collections[prefs.currentCategory].length()) {
                    prefs.currentToIndex = 0;
                }
                setUnitSpinners(prefs.currentCategory);
                prefs.save(WidgetSetupActivity.this);
            }

            @Override
            public void onNothingSelected(AdapterView<?> adapterView) {
            }
        });

        for (int i = 0; i < allCategoryNames.length; i++) {
            if (allCategoryNames[i].equals(collections[prefs.currentCategory].getNames()[0])) {
                unitCategorySpinner.setSelection(i);
            }
        }

        unitFromSpinner = (Spinner) findViewById(R.id.unit_from_spinner);
        unitFromSpinner.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
            @Override
            public void onItemSelected(AdapterView<?> adapterView, View view, int i, long l) {
                prefs.currentFromIndex = i;
                prefs.save(WidgetSetupActivity.this);
            }

            @Override
            public void onNothingSelected(AdapterView<?> adapterView) {
            }
        });

        unitToSpinner = (Spinner) findViewById(R.id.unit_to_spinner);
        unitToSpinner.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
            @Override
            public void onItemSelected(AdapterView<?> adapterView, View view, int i, long l) {
                prefs.currentToIndex = i;
                prefs.save(WidgetSetupActivity.this);
            }

            @Override
            public void onNothingSelected(AdapterView<?> adapterView) {
            }
        });
    }

    @Override
    public void onResume() {
        super.onResume();
        if (widgetId == -1) {
            finish();
        }
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item){
        switch(item.getItemId()) {
            case android.R.id.home:
                onBackPressed();
                return true;
        }
        return false;
    }

    private void setUnitSpinners(int category) {
        unitFromSpinner.setAdapter(new ArrayAdapter(this, R.layout.unit_categoryitem, collections[category].getItems()));
        unitFromSpinner.setSelection(prefs.currentFromIndex);

        unitToSpinner.setAdapter(new ArrayAdapter(this, R.layout.unit_categoryitem, collections[category].getItems()));
        unitToSpinner.setSelection(prefs.currentToIndex);
    }
}

File: app/src/main/java/com/defianttech/convertme/WidgetSetupActivity.kt
package com.defianttech.convertme

import android.os.Build
import android.os.Bundle
import androidx.appcompat.app.AppCompatActivity
import android.util.Log
import android.view.MenuItem
import android.view.View
import android.widget.AdapterView
import android.widget.ArrayAdapter
import com.defianttech.convertme.databinding.WidgetSetupActivityBinding

/*
 * Copyright (c) 2014-2020 Dmitry Brant
 */
class WidgetSetupActivity : AppCompatActivity() {
    private lateinit var binding: WidgetSetupActivityBinding

    private var collections = UnitCollection.getInstance(this)
    private var allCategoryNames = UnitCollection.getAllCategoryNames(this)
    private var widgetId = -1
    private lateinit var prefs: WidgetPrefs

    public override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = WidgetSetupActivityBinding.inflate(layoutInflater)
        setContentView(binding.root)
        setSupportActionBar(binding.toolbar)
        supportActionBar!!.setTitle(R.string.configure_widget)

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            window.navigationBarColor = getColor(R.color.number_pad_background)
        }

        if (!intent.action.isNullOrEmpty() && intent.action!!.contains(WidgetProvider.CLICK_ACTION_SETTINGS)) {
            widgetId = WidgetProvider.getWidgetId(intent.action!!)
        }
        if (widgetId == -1) {
            return
        }
        Log.d("WidgetSetupActivity", "Configuring widget $widgetId")

        prefs = WidgetPrefs(this, widgetId)

        val categoryAdapter = ArrayAdapter(this, R.layout.unit_categoryitem, allCategoryNames)

        binding.unitCategorySpinner.adapter = categoryAdapter
        binding.unitCategorySpinner.onItemSelectedListener = object : AdapterView.OnItemSelectedListener {
            override fun onItemSelected(adapterView: AdapterView<*>, view: View, i: Int, l: Long) {
                prefs.currentCategory = UnitCollection.collectionIndexByName(collections, allCategoryNames[i])
                if (prefs.currentFromIndex > collections[prefs.currentCategory].length()) {
                    prefs.currentFromIndex = 0
                }
                if (prefs.currentToIndex > collections[prefs.currentCategory].length()) {
                    prefs.currentToIndex = 0
                }
                setUnitSpinners(prefs.currentCategory)
                prefs.save(this@WidgetSetupActivity)
            }

            override fun onNothingSelected(adapterView: AdapterView<*>) {}
        }

        for (i in allCategoryNames.indices) {
            if (allCategoryNames[i] == collections[prefs.currentCategory].names[0]) {
                binding.unitCategorySpinner.setSelection(i)
            }
        }

        binding.unitFromSpinner.onItemSelectedListener = object : AdapterView.OnItemSelectedListener {
            override fun onItemSelected(adapterView: AdapterView<*>, view: View, i: Int, l: Long) {
                prefs.currentFromIndex = i
                prefs.save(this@WidgetSetupActivity)
            }

            override fun onNothingSelected(adapterView: AdapterView<*>) {}
        }

        binding.unitToSpinner.onItemSelectedListener = object : AdapterView.OnItemSelectedListener {
            override fun onItemSelected(adapterView: AdapterView<*>, view: View, i: Int, l: Long) {
                prefs.currentToIndex = i
                prefs.save(this@WidgetSetupActivity)
            }

            override fun onNothingSelected(adapterView: AdapterView<*>) {}
        }

        binding.unitIncrementText.setText(prefs.increment.toString())
    }

    public override fun onResume() {
        super.onResume()
        if (widgetId == -1) {
            finish()
        }
    }

    public override fun onPause() {
        super.onPause()
        try {
            prefs.increment = java.lang.Float.parseFloat(binding.unitIncrementText.text.toString())
        } catch (e: NumberFormatException) {
            prefs.increment = 1f
        }

        prefs.save(this)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            android.R.id.home -> {
                onBackPressedDispatcher.onBackPressed()
                return true
            }
        }
        return false
    }

    private fun setUnitSpinners(category: Int) {
        binding.unitFromSpinner.adapter = ArrayAdapter(this, R.layout.unit_categoryitem, collections[category].items)
        binding.unitFromSpinner.setSelection(prefs.currentFromIndex)

        binding.unitToSpinner.adapter = ArrayAdapter(this, R.layout.unit_categoryitem, collections[category].items)
        binding.unitToSpinner.setSelection(prefs.currentToIndex)
    }
}
